src/app/sidebar/sidebar/sidebar.component.ts

```diff
@@ -114,38 +120,159 @@ export class SidebarComponent {
   async downloadMap(map: ZsMapStateSource) {
     this.mapDownloadStates[map] = 'loading';
     const downloadUrl = zsMapStateSourceToDownloadUrl[map];
-    let localMapMeta = await db.localMapMeta.get(downloadUrl);
-    if (!localMapMeta) {
-      localMapMeta = {
-        url: downloadUrl,
-        map,
-        mapStatus: this.mapDownloadStates[map],
-        objectUrl: undefined,
-        mapStyle: undefined,
-      };
-      await db.localMapMeta.put(localMapMeta);
-    }
+    const localMapMeta = (await db.localMapMeta.get(downloadUrl)) || {
+      url: downloadUrl,
+      map,
+      mapStatus: 'missing',
+      objectUrl: undefined,
+      mapStyle: undefined,
+    };
+
+    await db.localMapMeta.put(localMapMeta);
+
     let localMapBlob = await db.localMapBlobs.get(localMapMeta.url);
     if (!localMapBlob) {
-      try {
-        const [localMap, mapStyle] = await Promise.all([
-          lastValueFrom(this.http.get(downloadUrl, { responseType: 'blob' })),
-          fetch('/assets/map-style.json').then((res) => res.text()),
-        ]);
-        localMapBlob = {
-          url: localMapMeta.url,
-          data: localMap,
-        };
-        await db.localMapBlobs.add(localMapBlob);
-        localMapMeta.mapStyle = mapStyle;
-        localMapMeta.objectUrl = undefined;
-        localMapMeta.mapStatus = 'downloaded';
-      } catch (e) {
+      const mapRequest = this.http
+        .get(downloadUrl, {
+          responseType: 'blob',
+          reportProgress: true,
+          observe: 'events',
+        })
+        .subscribe({
+          next: async (event) => {
+            if (event.type === HttpEventType.DownloadProgress) {
+              const percentDone = Math.round((100 * event.loaded) / (event.total ?? 1));
+              this.mapProgress = percentDone;
+              this.cdRef.detectChanges();
+            } else if (event instanceof HttpResponse) {
+              const localMap = event.body as Blob;
+              const mapStyle = await fetch('/assets/map-style.json').then((res) => res.text());
+              localMapBlob = {
+                url: localMapMeta.url,
+                data: localMap,
+              };
+              await db.localMapBlobs.add(localMapBlob);
+              localMapMeta.mapStyle = mapStyle;
+              localMapMeta.objectUrl = undefined;
+              localMapMeta.mapStatus = 'downloaded';
+              this.mapDownloadStates[map] = localMapMeta.mapStatus;
+              await db.localMapMeta.put(localMapMeta);
+              this.cdRef.detectChanges();
```

Comment:
also unsubscribe here?
```suggestion
              this.cdRef.detectChanges();
              this.subscriptions.get(map)?.unsubscribe();
```